File: stvs/src/main/java/edu/kit/iti/formal/stvs/model/expressions/VariableExpr.java
package edu.kit.iti.formal.stvs.model.expressions;

import edu.kit.iti.formal.stvs.model.common.FreeVariable;
import edu.kit.iti.formal.stvs.model.common.IoVariable;

import java.util.Optional;
import java.util.regex.Pattern;

/**
 * Runtime-representation for variables in {@link Expression}s.
 * At this point it is not known whether this is a reference to a {@link FreeVariable} or an
 * {@link IoVariable}; it is simply the String name of either of those.
 *
 * @author Philipp
 */
public class VariableExpr extends Expression {

  public static final Pattern IDENTIFIER_PATTERN = Pattern.compile("[a-zA-Z_][$a-zA-Z0-9_]*");

  private final String varName;
  private final Optional<Integer> index;

  /**
   * Constructs a new VariableExpr with a backwards reference.
   *
   * @param varName the name as a reference to a variable
   * @param index the index of the backwards-reference (for expressions like <tt>A[-1]</tt> for
   *        example)
   */
  public VariableExpr(String varName, int index) {
    this.varName = varName;
    this.index = Optional.of(index);
  }

  /**
   * Constructs a new VariableExpr without a backwards reference.
   *
   * @param name the name as a reference to a variable.
   */
  public VariableExpr(String name) {
    this.varName = name;
    this.index = Optional.empty();
  }

  @Override
  public <R> R takeVisitor(ExpressionVisitor<R> visitor) {
    return visitor.visitVariable(this);
  }

  public String getVariableName() {
    return varName;
  }

  public Optional<Integer> getIndex() {
    return index;
  }

  public boolean equals(VariableExpr expr) {
    return getVariableName().equals(expr.getVariableName());
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }

    VariableExpr that = (VariableExpr) o;

    if (varName != null ? !varName.equals(that.varName) : that.varName != null) {
      return false;
    }
    return getIndex() != null ? getIndex().equals(that.getIndex()) : that.getIndex() == null;
  }

  @Override
  public int hashCode() {
    int result = varName != null ? varName.hashCode() : 0;
    result = 31 * result + (getIndex() != null ? getIndex().hashCode() : 0);
    return result;
  }

  public String toString() {
    String indexStr = index.map(i -> "[" + i + "]").orElse("");
    return "VariableExpr(" + varName + indexStr + ")";
  }

}

File: stvs/src/main/kotlin/edu/kit/iti/formal/stvs/model/expressions/VariableExpr.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.model.expressions

import edu.kit.iti.formal.stvs.model.common.FreeVariable
import edu.kit.iti.formal.stvs.model.common.IoVariable
import java.util.regex.Pattern

/**
 * Runtime-representation for variables in [Expression]s.
 * At this point it is not known whether this is a reference to a [FreeVariable] or an
 * [IoVariable]; it is simply the String name of either of those.
 *
 * @author Philipp
 */
data class VariableExpr(val variableName: String?, val index: Int? = null) : Expression() {
    override fun <R> accept(visitor: ExpressionVisitor<R>): R = visitor.visitVariable(this)

    fun equals(expr: VariableExpr): Boolean = variableName == expr.variableName

    override fun toString(): String {
        val indexStr = index?.let { "[$it]" } ?: ""
        return "VariableExpr($variableName$indexStr)"
    }

    companion object {
        @JvmField
        val IDENTIFIER_PATTERN: Pattern = Pattern.compile("[a-zA-Z_][\$a-zA-Z0-9_]*")
    }
}